### PR TITLE
Add support for functions on Enum.with_index/2

### PR DESCRIPTION
**Pull request by v0idpwn**
Makes  `Enum.with_index/2` support both functions and offsets by setting
a default function `fn element, index -> {element, index} end` and
a default offset `0` and letting the user change one of them in the
second argument.

### CODE REVIEW

**Review by Eiji7 — COMMENTED** [review]
Looks nice. I have added suggestions in case we want to be more strict.

**Comment by v0idpwn** [comment]
@Eiji7 thanks for taking your time to review! 

The support for functions should allow you to use the indexes however you want. For instance, you could change the tuple order (scenario which motivated this PR, see https://groups.google.com/g/elixir-lang-core/c/lZyKHBJOVHc/), this implies that the result spec could be any list()

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -3226,7 +3226,13 @@ defmodule Enum do
   Returns the `enumerable` with each element wrapped in a tuple
   alongside its index.
 
-  If an `offset` is given, we will index from the given offset instead of from zero.
+  May receive a function or an offset.
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex:3229]:
```suggestion
  May receive a function or an integer offset.
```

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -3236,21 +3242,31 @@ defmodule Enum do
...
+  @spec do_with_index(list, integer, function) :: list
+  defp do_with_index([], _offset, _fun) do
     []
   end
 
-  defp do_with_index([head | tail], index) do
-    [{head, index} | do_with_index(tail, index + 1)]
+  defp do_with_index([head | tail], index, fun) do
+    [fun.(head, index) | do_with_index(tail, index + 1, fun)]
   end
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex:3270]: The implementation we have here is not very efficient unfortunately. For things that are not lists, it traverses everything twice. I think we should write it like this:

```elixir
def with_index(enumerable, offset) when is_integer(offset) do
  enumerable
  |> Enumerable.map_reduce(offset, fn x, i -> {{x, i}, i+1} end)
  |> elem(0)
end

def with_index(enumerable, fun) when is_function(fun) do
  enumerable
  |> Enumerable.map_reduce(offset, fn x, i -> {fun.(x, i), i+1} end)
  |> elem(0)
end
```

Can you please update accordingly?

Thank you.

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -3236,21 +3242,31 @@ defmodule Enum do
       iex> Enum.with_index([:a, :b, :c], 3)
       [a: 3, b: 4, c: 5]
 
+      iex> Enum.with_index([:a, :b, :c], fn element, index -> {index, element} end)
+      [{0, :a}, {1, :b}, {2, :c}]
   """
-  @spec with_index(t, integer) :: [{element, index}]
-  def with_index(enumerable, offset \\ 0) do
+  @spec with_index(t, integer | (element, index -> any)) :: list
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex:3248]:
```suggestion
  @spec with_index(t, integer) :: [{term, integer}]
  @spec with_index(t, (element, index -> value)) :: [value] when value: any
```

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (v0idpwn) [reply lib/elixir/lib/enum.ex:3270]: Sure!

File: lib/elixir/lib/enum.ex (hunk 3 above)
Comment (v0idpwn) [reply lib/elixir/lib/enum.ex:3248]: Didn't even knew you could write specs this way T_T

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (v0idpwn) [reply lib/elixir/lib/enum.ex]: Just pushed the updated version, @josevalim :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**